ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1651,7 +1651,7 @@ private InstallSnapshotReplyProto notifyStateMachineToInstallSnapshot(
       }
       changeToFollowerAndPersistMetadata(leaderTerm, "installSnapshot");
       state.setLeader(leaderId, "installSnapshot");
-      long snapshotIndex = state.getSnapshotIndex();
+      long snapshotIndex = state.getLog().getSnapshotIndex();
```

Comment:
Why is this change required? ServerState snapshotIndex should be the same as RaftLog snapshotIndex. 
Please let me know if I am missing something here.

Reply:
There is the possibility that after statemachine just installed the snapshot and before the main thread return the  SNAPSHOT_INSTALLED, the snapshotIndex here will get from the statemachine instead of memory and  return ALREADY_INSTALLED, which will cause the follower not to update the index and reload statemachine

Reply:
> There is the possibility that after statemachine just installed the snapshot and before the main thread return the SNAPSHOT_INSTALLED, the snapshotIndex here will get from the statemachine instead of memory and return ALREADY_INSTALLED, which will cause the follower not to update the index and reload statemachine

before the main thread return the SNAPSHOT_INSTALLED, `inProgressInstallSnapshotRequest` will not be set back to 0 if no exception is thrown. so the following code will not be executed until `inProgressInstallSnapshotRequest` is set back to 0. 
```
        if (snapshotIndex + 1 >= firstAvailableLogIndex && firstAvailableLogIndex > 0) {
          // State Machine has already installed the snapshot. Return the
          // latest snapshot index to the Leader.

          inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogIndex, 0);
          LOG.info("{}: InstallSnapshot notification result: {}, current snapshot index: {}", getMemberId(),
              InstallSnapshotResult.ALREADY_INSTALLED, snapshotIndex);
          return ServerProtoUtils.toInstallSnapshotReplyProto(leaderId, getMemberId(), currentTerm,
              InstallSnapshotResult.ALREADY_INSTALLED, snapshotIndex);
        }
```
if `inProgressInstallSnapshotRequest` is set back to 0, it means the snapshot has been installed successfully,  so now ServerState snapshotIndex should be the same as RaftLog snapshotIndex because the statemachine has been reloaded. 
so seems the changes here is unnecessary , please correct me if i miss something here!